python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,1111 @@
+            self.node_config["library_item"] = freeze_vm["library_item"]
+            self.node_config["resource_pool"] = freeze_vm["resource_pool"]
+            self.node_config["resources"] = freeze_vm["resources"]
+            self.node_config["networks"] = freeze_vm["networks"]
+            self.node_config["keep_nics_disconnected"] = True
+            self.node_config["datastore"] = freeze_vm["datastore"]
+            self.outer_obj = outer_obj
+            self.target_name = target_name
+            self.thread_exception = None
+            super().__init__(args=(target_name, node_config, outer_obj))
+
+        def run(self):
+            try:
+                vm_name = self.target_name
+                vms = self.outer_obj.vsphere_automation_sdk_client.vcenter.VM.list(
+                    VM.FilterSpec(names=set([vm_name]))
+                )
+
+                # Frozen VM already present so no need to create again
+                if len(vms) > 0:
+                    cli_logger.info("Frozen VM already exists. Not creating again.")
+
+                    vm_id = vms[0].vm
+                    status = self.vsphere_automation_sdk_client.vcenter.vm.Power.get(
+                        vm_id
+                    )
+
+                    if status.state != HardPower.State.POWERED_ON:
+                        cli_logger.info("Frozen VM is off. Powering it ON")
+                        self.vsphere_automation_sdk_client.vcenter.vm.Power.start(vm_id)
+                        cli_logger.info("vm.Power.start({})".format(vm_id))
+                    return
+
+                cli_logger.info("Creating frozen VM")
+                self._return = self.outer_obj.create_ovf_node(self.node_config, vm_name)
+            except Exception as e:
+                cli_logger.info("Failed creating frozen VM {}".format(self.target_name))
+                self.thread_exception = e
+
+        def join(self):
+            return super().join()
+
+    @staticmethod
+    def bootstrap_config(cluster_config):
+        return bootstrap_vsphere(cluster_config)
+
+    def non_terminated_nodes(self, tag_filters):
+        """For vSphere, non_terminated nodes are:
+        1. Nodes in creating state.
+        2. Nodes already created and powered on.
+        """
+        with self.lock:
+            nodes = []
+            cli_logger.info("Getting non terminated nodes...")
+            vms = self.vsphere_automation_sdk_client.vcenter.VM.list()
+            cli_logger.info(f"Got {len(vms)} non terminated nodes.")
+            filters = tag_filters.copy()
+            if TAG_RAY_CLUSTER_NAME not in tag_filters:
+                filters[TAG_RAY_CLUSTER_NAME] = self.cluster_name
+
+            for vm in vms:
+                vm_id = vm.vm
+                yn_id = DynamicID(type=Constants.TYPE_OF_RESOURCE, id=vm.vm)
+
+                matched_tags = {}
+                all_tags = {}
+                # If the VM has a tag from tag_filter then select it
```

Comment:
We mean to select it if it has *all* tags from the tag_filter, not just *a* tag, right? It looks like that's what the code is doing

Reply:
Addressed